src/engine/TransitivePathImpl.h

```diff
@@ -190,122 +195,124 @@ class TransitivePathImpl : public TransitivePathBase {
    *
    * @param edges Adjacency lists, mapping Ids (nodes) to their connected
    * Ids.
-   * @param nodes A list of Ids. These Ids are used as starting points for the
-   * transitive hull. Thus, this parameter guides the performance of this
-   * algorithm.
+   * @param startNodes A range that yields an instanciation of
+   * `TableColumnWithVocab` that can be consumed to create a transitive hull.
    * @param target Optional target Id. If supplied, only paths which end
    * in this Id are added to the hull.
    * @return Map Maps each Id to its connected Ids in the transitive hull
    */
-  Map transitiveHull(const T& edges, const std::vector<Id>& startNodes,
-                     std::optional<Id> target) const {
-    // For every node do a dfs on the graph
-    Map hull{allocator()};
-
+  NodeGenerator transitiveHull(const T& edges, LocalVocab edgesVocab,
+                               std::ranges::range auto startNodes,
+                               std::optional<Id> target) const {
+    ad_utility::Timer timer{ad_utility::Timer::Stopped};
     std::vector<std::pair<Id, size_t>> stack;
     ad_utility::HashSetWithMemoryLimit<Id> marks{
         getExecutionContext()->getAllocator()};
-    for (auto startNode : startNodes) {
-      if (hull.contains(startNode)) {
-        // We have already computed the hull for this node
-        continue;
-      }
-
-      marks.clear();
-      stack.clear();
-      stack.push_back({startNode, 0});
-
-      if (minDist_ == 0 &&
-          (!target.has_value() || startNode == target.value())) {
-        insertIntoMap(hull, startNode, startNode);
-      }
-
-      while (!stack.empty()) {
-        checkCancellation();
-        auto [node, steps] = stack.back();
-        stack.pop_back();
+    for (auto&& tableColumn : startNodes) {
+      timer.cont();
+      LocalVocab mergedVocab = std::move(tableColumn.vocab_);
+      mergedVocab.mergeWith(std::span{&edgesVocab, 1});
+      size_t currentRow = 0;
+      for (Id startNode : tableColumn.column_) {
+        Set connectedNodes{getExecutionContext()->getAllocator()};
```

Comment:
The body of this loop
can and probably should be a separate function (computethe hull for a single IdTable), else the nesting level is a little deep.